### PR TITLE
Support Layer-wise Adaptive Rate Scaling(LARS)

### PR DESCRIPTION
**Pull request by tohmae**
This PR is proposal for supporting a new optimizer-hook function based on Layer-wise Adaptive Rate Scaling.
Deteil is here.
https://arxiv.org/abs/1708.03888
https://arxiv.org/abs/1801.03137

Usage is as follows.
optimizer = optimizers.SGD(lr=0.01)
or
optimizer = optimizers.MomentumSGD(lr=0.01, momentum=0.9)
optimizer.setup(model)
optimizer.add_hook(chainer.optimizer.GradientLARS(weightdecay=0.0005))
lr(in chainer) = γ * η(in paper)

### CODE REVIEW

**Comment by tohmae** [comment]
Please advise me.  😢

File: chainer/optimizer.py [hunk 1]
```diff
@@ -853,3 +853,58 @@ def __call__(self, rule, param):
...
+
+    See: `Large Batch Training of Convolutional Networks \
+          <https://arxiv.org/abs/1708.03888>`_
+
+    This hook function scales all gradient arrays to fit to the weight norm.
+
+    Args:
+        threashold (float): If weight norm is more than threshold,
+            this function scales all gradient arrays to fit weight norm.
+        weightdecay (float): Coefficient for the weight decay.
```
Comment (delta2323) [inline chainer/optimizer.py]: The WeightDecay hook treat the word "weightdecay" as a concatenation of two words. To align with the convention, could you write as `weight_decay`?

File: chainer/optimizer.py [hunk 2]
```diff
@@ -853,3 +853,58 @@ def __call__(self, rule, param):
...
+        eps (float): Small value for the numerical stability.
+
+    Attributes:
+        threashold (float): If weight norm is more than threshold,
+            this function scales all gradient arrays to fit weight norm.
+        weightdecay (float): Coefficient for the weight decay.
```
Comment (delta2323) [inline chainer/optimizer.py]: ditto (write as `weight_decay`).

File: chainer/optimizer.py [hunk 3]
```diff
@@ -853,3 +853,58 @@ def __call__(self, rule, param):
...
+        eps (float): Small value for the numerical stability.
+    """
+    name = 'GradientLARS'
+    call_for_each_param = True
+
+    def __init__(self, threshold=1e-2, weightdecay=0.0, eps=1e-9):
```
Comment (delta2323) [inline chainer/optimizer.py]: Could you let me know how did you determine the `eps` value. I could not find it in the papers (correct me if I miss it).

File: chainer/optimizer.py (hunk 1 above)
Comment (tohmae) [reply chainer/optimizer.py]: OK.

File: chainer/optimizer.py (hunk 2 above)
Comment (tohmae) [reply chainer/optimizer.py]: OK.

File: chainer/optimizer.py (hunk 3 above)
Comment (tohmae) [reply chainer/optimizer.py]: I refer to source code in  https://github.com/bparr/lars written by tensorflow.

**Comment by tohmae** [comment]
Thank you for your comment. 😄

File: chainer/optimizer.py (hunk 3 above)
Comment (delta2323) [reply chainer/optimizer.py]: OK, then please write in the doc string that your implementation is the version the paper (https://arxiv.org/abs/1801.03137) and introduced eps term, as described in section 6.1 of the paper.

**Comment by delta2323** [comment]
Please refer https://arxiv.org/abs/1801.03137 (as well as https://arxiv.org/abs/1708.03888) because implementation is based on that paper.

File: chainer/optimizer.py [hunk 4]
```diff
@@ -853,3 +853,58 @@ def __call__(self, rule, param):
...
+            return
+
+        xp = cuda.get_array_module(p)
+
+        # weight norm
+        p_norm = xp.linalg.norm(p)
+        # grad norm
+        g_norm = xp.linalg.norm(g)
+        local_rate = p_norm / (self.eps + g_norm + self.weight_decay * p_norm)
+        rate = xp.where(p_norm > self.threshold, local_rate, 1.0)
```
Comment (delta2323) [inline chainer/optimizer.py]: Please write this thresholding comes from [this implementation](https://github.com/bparr/lars/blob/master/lars.py#L77), as it is not written in https://arxiv.org/abs/1708.03888 nor https://arxiv.org/abs/1801.03137 (if I do not miss some information).

File: chainer/optimizer.py [hunk 5]
```diff
@@ -853,3 +853,58 @@ def __call__(self, rule, param):
...
+        if p is None or g is None:
+            return
+
+        xp = cuda.get_array_module(p)
+
+        # weight norm
+        p_norm = xp.linalg.norm(p)
+        # grad norm
+        g_norm = xp.linalg.norm(g)
+        local_rate = p_norm / (self.eps + g_norm + self.weight_decay * p_norm)
```
Comment (delta2323) [inline chainer/optimizer.py]: It could be confusing to call this value `local_rate` because local LR (learning rate) λ^l is defined as multiplication of η (LARS coefficient) and this value in (6) of https://arxiv.org/pdf/1708.03888.pdf.

File: chainer/optimizer.py [hunk 6]
```diff
@@ -853,3 +853,58 @@ def __call__(self, rule, param):
...
         with cuda.get_device_from_array(grad):
             xp.clip(grad, self.lower_bound, self.upper_bound, out=grad)
+
+
+class GradientLARS(object):
+
+    """Optimizer/UpdateRule hook function for layer wise adaptive rate scaling.
+
+    See: `Large Batch Training of Convolutional Networks \
+          <https://arxiv.org/abs/1708.03888>`_
```
[the remark this answers is not in the document]
Comment (tohmae) [reply chainer/optimizer.py]: OK.

File: chainer/optimizer.py (hunk 4 above)
Comment (tohmae) [reply chainer/optimizer.py]: OK.

File: chainer/optimizer.py (hunk 5 above)
Comment (tohmae) [reply chainer/optimizer.py]: OK.

**Comment by tohmae** [comment]
Thank you for your comment. 
To avoid confusion I add many comments.
Please check them.

File: chainer/optimizer_hooks/gradient_lars.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+       : `Convergence Analysis of Gradient Descent Algorithms \
+          with Proportional Updates \
+          <https://arxiv.org/abs/1801.03137>`_
+
+    This hook function scales all gradient arrays to fit to the weight norm.
+
+    In <https://arxiv.org/abs/1708.03888>
+        \\Delta w_{t+1} = m * \\Delta w_t
+                        + \\gamma * \\lambda * (\\nabla L_t(w) + \\beta w_t)
+        where
```
[the remark this answers is not in the document]
Comment (delta2323) [reply chainer/optimizer_hooks/gradient_lars.py]: `lambda` is a time-dependent variable as it is a function of the current weight `w_t`. But so strictly speaking, it may be better to write as `lambda_t`. But current document is still OK because the original paper (https://arxiv.org/abs/1708.03888) also write this way.

File: chainer/optimizer_hooks/gradient_lars.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+           \\gamma : learning_rate
+           m       : momentum
+           \\beta  : weight_decay
+           \\eta   : lars_coeeficient
+           \\lamda : local_lr
+                     = \\eta * \\frac{\|w\|}{\|\\nabla L(w)\| + \\beta * \|w\|}
```
Comment (delta2323) [inline chainer/optimizer_hooks/gradient_lars.py]: I think replacing `w` with `w_t` would be good as the unsubscripted `w` is not used in the previous fomula.

File: chainer/optimizer_hooks/gradient_lars.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+    See: `Large Batch Training of Convolutional Networks \
+          <https://arxiv.org/abs/1708.03888>`_
+       : `Convergence Analysis of Gradient Descent Algorithms \
+          with Proportional Updates \
+          <https://arxiv.org/abs/1801.03137>`_
+
+    This hook function scales all gradient arrays to fit to the weight norm.
+
+    In <https://arxiv.org/abs/1708.03888>
```
Comment (delta2323) [inline chainer/optimizer_hooks/gradient_lars.py]: Add a comma (',') after the URL.

File: chainer/optimizer_hooks/gradient_lars.py (hunk 8 above)
Comment (delta2323) [reply chainer/optimizer_hooks/gradient_lars.py]: Add a period (`.`) after the last formula.

File: chainer/optimizer_hooks/gradient_lars.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+        \\Delta w_{t+1} = m * \\Delta w_t
+                        + \\gamma * \\lambda * (\\nabla L_t(w) + \\beta w_t)
+        where
+           \\gamma : learning_rate
+           m       : momentum
+           \\beta  : weight_decay
+           \\eta   : lars_coeeficient
+           \\lamda : local_lr
+                     = \\eta * \\frac{\|w\|}{\|\\nabla L(w)\| + \\beta * \|w\|}
+    In GradientLARS
```
Comment (delta2323) [inline chainer/optimizer_hooks/gradient_lars.py]: ditto (add a comma)

File: chainer/optimizer_hooks/gradient_lars.py [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+        lr : lr in chainer.optimizers.SGD or chainer.optimizers.MomentumSGD
+        clip_rate : \\frac{\|w\|}{\|\\nabla L(w)\| + \\beta * \|w\|}
```
Comment (delta2323) [inline chainer/optimizer_hooks/gradient_lars.py]: dotto (add a period)

File: chainer/optimizer_hooks/gradient_lars.py [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    Then
+        \\gamma * \\lamda in arXiv = lr * clip_rate in GradientLARS and
+        \\gamma * \\eta in arXiv = lr in GradientLARS
+    So
```
Comment (delta2323) [inline chainer/optimizer_hooks/gradient_lars.py]: ditto (add a comma)

File: chainer/optimizer_hooks/gradient_lars.py [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+           \\beta  : weight_decay
+           \\eta   : lars_coeeficient
+           \\lamda : local_lr
+                     = \\eta * \\frac{\|w\|}{\|\\nabla L(w)\| + \\beta * \|w\|}
+    In GradientLARS
+        lr : lr in chainer.optimizers.SGD or chainer.optimizers.MomentumSGD
+        clip_rate : \\frac{\|w\|}{\|\\nabla L(w)\| + \\beta * \|w\|}
+    Then
+        \\gamma * \\lamda in arXiv = lr * clip_rate in GradientLARS and
+        \\gamma * \\eta in arXiv = lr in GradientLARS
```
Comment (delta2323) [inline chainer/optimizer_hooks/gradient_lars.py]: ditto (add a period)

File: chainer/optimizer_hooks/gradient_lars.py [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+           m       : momentum
+           \\beta  : weight_decay
+           \\eta   : lars_coeeficient
+           \\lamda : local_lr
+                     = \\eta * \\frac{\|w\|}{\|\\nabla L(w)\| + \\beta * \|w\|}
+    In GradientLARS
+        lr : lr in chainer.optimizers.SGD or chainer.optimizers.MomentumSGD
+        clip_rate : \\frac{\|w\|}{\|\\nabla L(w)\| + \\beta * \|w\|}
+    Then
+        \\gamma * \\lamda in arXiv = lr * clip_rate in GradientLARS and
```
Comment (delta2323) [inline chainer/optimizer_hooks/gradient_lars.py]: IMO, it would be less obvious where does this formula come from. So I would suggest, some thing like this:

As `lr` corresponds to `lambda * eta`, we define `clip_rate` as `\frac{|w|}{\|nabla L(w)| + \beta * |w|}` and reformulate the aforementioned formula as:
`\Delta w_{t+1} = m * \Delta w_t + lr * clip_rate * (\nabla L_t(w) + \beta w_t)`
and implement in this way.

**Comment by delta2323** [comment]
Could you rebase the latest master branch?

**Comment by tohmae** [comment]
Thank you for your comments. Please review again.

File: chainer/optimizer_hooks/gradient_lars.py [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+    See: `Large Batch Training of Convolutional Networks \
+          <https://arxiv.org/abs/1708.03888>`_
+       : `Convergence Analysis of Gradient Descent Algorithms \
+          with Proportional Updates \
+          <https://arxiv.org/abs/1801.03137>`_
+
+    This hook function scales all gradient arrays to fit to the weight norm.
+
+    In <https://arxiv.org/abs/1708.03888>,
+        v_{t+1} = m * v_t + \\gamma * \\lambda * (\\nabla L(w_t) + \\beta w_t)
```
Comment (delta2323) [inline chainer/optimizer_hooks/gradient_lars.py]: Could you use `.. math::` directive ([document](http://www.sphinx-doc.org/en/master/ext/math.html#directive-math), [example](https://github.com/chainer/chainer/blob/3ba7d4d0ddcc0202225161f75e6d6fed07ffa25e/chainer/functions/noise/gumbel_softmax.py))?

Comment (delta2323) [reply chainer/optimizer_hooks/gradient_lars.py]: You can check if the directive works by compiling the document. We can compile the document by typing 
`sphinx-build -b html -d build/doctrees source build/html` at `docs` directory in the Chainer repository.

**Comment by delta2323** [comment]
Could you add `GradientLARS` here so that this module is available in the document.

**Comment by tohmae** [comment]
Thank you for your comments. I modify comment in GradientLARS so that this module is available in the document. Please review again.

**Comment by delta2323** [comment]
jenkins, test this please.

**Comment by delta2323** [comment]
LGTM, I launched CI to for final check.

**Comment by delta2323** [comment]
The following warning occured in testing the document in the CI.

```
13:16:43 /home/user/.local/lib/python3.5/site-packages/chainer-4.0.0rc1-py3.5.egg/chainer/optimizer_hooks/gradient_lars.py:docstring of chainer.optimizer_hooks.GradientLARS:None: WARNING: more than one target found for cross-reference 'eps': chainer.optimizers.SMORMS3.eps, chainer.optimizers.AdaGrad.eps, chainer.optimizers.Adam.eps, chainer.optimizers.RMSpropGraves.eps, chainer.optimizers.RMSprop.eps, chainer.optimizers.AdaDelta.eps
```

As we treat a warning in generating the document as an error, so I find a way to resolve this warning. If you can do it by yourself, please fix the code.

**Comment by delta2323** [comment]
@tohmae Thank you! I'll run the CI again.

**Comment by delta2323** [comment]
jenkins, test this please.

File: chainer/optimizer_hooks/gradient_lars.py [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+        ~optimizer_hooks.GradientLARS.eps (float): Small value for the
+                         numerical stability.
+                         (See <https://arxiv.org/abs/1801.03137>)
+        ~optimizer_hooks.GradientLARS.timing (string): Specifies
+                         when this hook should be called by the
+                         Optimizer/UpdateRule. Valid values are 'pre'
+                         (before any updates) and 'post' (after any updates).
+
+    .. versionadded:: 4.0.0
+       The *timing* parameter.
```
Comment (delta2323) [inline chainer/optimizer_hooks/gradient_lars.py]: You do not need these lines as this hook itself will be added after v4.

**Comment by tohmae** [comment]
Thank you for your comment.

**Comment by delta2323** [comment]
LGTM
